src/pudl/transform/ferc714.py

```diff
@@ -409,139 +402,385 @@ def core_ferc714__respondent_id(
-    df = df.drop(columns="utc_offset_code")
+        merge_keys = ["entity_id", "report_year", "report_day", "sched_table_name"]
+        # Merge instant into duration.
+        out_df = pd.merge(
+            instant,
+            duration,
+            how="left",
+            on=merge_keys,
+            validate="m:1",
+        ).drop(columns=["report_day", "start_date", "end_date"])
+        return out_df
+
+    @staticmethod
+    def convert_dates_to_zero_offset_hours_xbrl(xbrl: pd.DataFrame) -> pd.DataFrame:
+        """Convert all hours to: Hour (24-hour clock) as a zero-padded decimal number.
+
+        Some but not all of the records start with hour 0, while other start with hour 1.
+        It is not immediately clear whether or not hours 1-24 corresponds to 1-00 hours.
+        """
+        bad_24_hour_mask = xbrl.report_date.str.contains("T24:")
+
+        xbrl.loc[bad_24_hour_mask, "report_date"] = pd.to_datetime(
+            xbrl[bad_24_hour_mask].report_date.str.replace("T24:", "T23:"),
+            format="%Y-%m-%dT%H:%M:%S",
+        ) + np.timedelta64(1, "h")
+        return xbrl
+
+    @staticmethod
+    def convert_dates_to_zero_seconds_xbrl(xbrl: pd.DataFrame) -> pd.DataFrame:
+        """Convert the last second of the day records to the first (0) second of the next day.
+
+        There are a small amount of records which report the last "hour" of the day
+        with as last second of the day, as opposed to T24 cleaned in
+        :func:`convert_dates_to_zero_offset_hours_xbrl` or T00 which is standard for a
+        numpy datetime. This function finds these records and adds one second of them and
+        then ensures all of the records has 0's for seconds.
+        """
+        last_second_mask = xbrl.report_date.dt.second == 59
+
+        xbrl.loc[last_second_mask, "report_date"] = xbrl.loc[
+            last_second_mask, "report_date"
+        ] + pd.Timedelta("1s")
+        assert xbrl[xbrl.report_date.dt.second != 0].empty
+        return xbrl
+
+    @staticmethod
+    def ensure_dates_are_complete_and_unique_xbrl(df):
+        """Assert that almost all respondents and years have complete and unique dates.
+
+        We found 41 gaps in the timeseries!
+        """
+        df["gap"] = df[["respondent_id_ferc714", "report_date"]].sort_values(
+            by=["respondent_id_ferc714", "report_date"]
+        ).groupby("respondent_id_ferc714").diff() > pd.to_timedelta("1h")
+        if len(gappy_dates := df[df.gap]) > 41:
+            raise AssertionError(
+                "We expect there to be nearly no gaps in the time series."
+                f"but we found these gaps:\n{gappy_dates}"
+            )
+        return df.drop(columns=["gap"])
+
+    @staticmethod
+    def parse_date_strings_xbrl(xbrl: pd.DataFrame) -> pd.DataFrame:
+        """Convert report_date into pandas Datetime types."""
+        xbrl = xbrl.astype({"report_date": "datetime64[ns]"})
+        return xbrl
```

Comment:
Why did you give this it's own function? Seems pretty small?

Reply:
yea i agree. previously this function was more like the `parse_date_strings_csv` in that it both converted dates and did the `ensure_dates_are_complete_and_unique_xbrl`. but i needed to do yet another time cleaning step before the ensure step so i broke them apart. I will just do this inline in the `run` method